src/photos/ducks/timeline-clusters/index.js

```diff
@@ -36,39 +38,122 @@ const TIMELINE_MUTATIONS = query => ({
     })
 })
 
-const getPhotosByMonth = photos => {
-  let sections = {}
-  photos.forEach(p => {
+const formatDate = date => {
+  return format(date, 'DD MMMM YYYY')
+}
+
+const sectionTitle = album => {
+  if (album.period) {
+    const startPeriod = new Date(album.period.start)
+    const endPeriod = new Date(album.period.end)
+    if (differenceInCalendarDays(endPeriod, startPeriod) > 0) {
+      // TODO Better period display
+      return formatDate(startPeriod) + ' - ' + formatDate(endPeriod)
+    }
+  }
+  return formatDate(new Date(album.name))
+}
+
+/**
+ * A section matches if:
+ * - The datetime is inside the period.
+ * - The section has no period (not a cluster) and is the datetime's day.
+ */
+const matchingSection = (sections, datetime) => {
+  return Object.keys(sections).find(date => {
+    if (sections[date].period) {
+      const startPeriod = new Date(sections[date].period.start)
+      const endPeriod = new Date(sections[date].period.end)
+      return datetime >= startPeriod && datetime <= endPeriod
+    } else {
+      // If the section has no period, it is not a cluster but a daily section
+      return differenceInCalendarDays(datetime, new Date(date)) === 0
+    }
+  })
+}
+
+/**
+ * Save the photos not clustered in the sections.
+ * It simulates a clustering, waiting for the actual one to be processed.
+ */
+const handlePhotosNotClustered = (sections, photosNotClustered) => {
+  photosNotClustered.forEach(photo => {
     const datetime =
-      p.metadata && p.metadata.datetime ? p.metadata.datetime : Date.now()
-    // here we want to get an object whose keys are months in a l10able format
-    // so we only keep the year and month part of the date
-    const month = datetime.slice(0, 7) + '-01T00:00'
-    /* istanbul ignore else */
-    if (!sections.hasOwnProperty(month)) {
-      sections[month] = []
+      photo.metadata && photo.metadata.datetime
+        ? photo.metadata.datetime
+        : photo.created_at
+
+    const sectionDate = matchingSection(sections, new Date(datetime))
+    if (sectionDate) {
+      sections[sectionDate].photos.push(photo)
+    } else {
+      // Create a new section for this day, without a period, to differentiate from clusters' sections
+      const day = datetime.slice(0, 10)
+      sections[day] = {
+        date: day,
+        title: formatDate(new Date(datetime)),
+        photos: [photo]
+      }
+    }
+  })
+}
+
+/**
+ * Retrieve photos by clusters. Each clusterized photo is referenced by an album
+ * having an 'auto' field to true.
+ * If a photo is not clusterized yet (newly updated ones), we try to insert them
+ * inside existing clusters, or group them per day otherwise.
+ */
+const getPhotosByClusters = photos => {
+  const sections = {}
+  const photosNotClustered = []
+
+  photos.forEach(p => {
+    const refAlbums = p.albums ? p.albums.data : null
+    // TODO Ensure unicity on the service side
+    const album = refAlbums ? refAlbums.find(ref => ref.auto) : null
```

Comment:
We can do things like this as well:

```

const refAlbums = p.albums ? p.albums.data : []
const album = refAlbums.find(ref => ref.auto)
```

And while we're at it we can probably simplify it some more:

```

const clusterAlbum = p.albums ? p.albums.data.find(ref => ref.auto) : null
```

What do you think?

Reply:
What is `p.albums.data` is `undefined`? Won't this fail with a `Cannot read property 'find' of undefined`?

Reply:
Ah shit yes, you're right. So maybe not the third version 😝